README.md
# mac-disk-monitor

Rust library and command-line tool to monitor disk activity on MacOS

[![CI](https://github.com/gabrielfalcao/mac-disk-monitor/actions/workflows/rust.yml/badge.svg)](https://github.com/gabrielfalcao/mac-disk-monitor/actions/workflows/rust.yml)
[![codecov](https://codecov.io/gh/gabrielfalcao/mac-disk-monitor/branch/main/graph/badge.svg?token=WjeDQXuLmv)](https://codecov.io/gh/gabrielfalcao/mac-disk-monitor)
![crates.io](https://img.shields.io/crates/v/mac-disk-monitor.svg)

## Usage

Run the code below, then try plugging a USB dongle or opening a DMG file.

```rust
use mac_disk_monitor::{stream_events, Action};
use std::sync::mpsc::channel;
use std::time::Duration;

fn main() {
    let (action, receiver) = channel();
    let (thread, receiver) = stream_events(receiver);

    loop {
        // poll for events every 1.5 seconds
        match receiver.recv_timeout(Duration::from_millis(1500)) {
            Ok(event) => match event {
                Some(event) => {
                    println!("{}", event.to_json());
                }
                None => {
                    eprintln!("thread has stopped...");
                }
            },
            Err(e) => {
                eprintln!("Error: {}", e);
                break;
            }
        }
    }
    action.send(Action::Stop).unwrap();
    eprintln!("waiting for thread to stop...");
    thread.join().unwrap().unwrap()
}
```

## Context

This started as a pet project to practice rust.

It executes the command `diskutil activity` in a thread and parses every stdout
line in real-time, extracting structured event data.

Developers should be able to subscribe to events and take action, for
example emitting a notification if a specific disk is mounted.


### Development

This repo is in active development with best effort to classic
Test-Driven Development, commits happen generaly in this order:

- "TDD [red] - write a failing test that given an input expects an output"
- "TDD [green] - write the smallest and/or simplest code to make the test pass, probably via hardcoding expected value"
- "TDD [refactor] - write real code to make the test pass

Other than that, all other commits kind of follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### Test Data

<details>
<summary>Click here to see the test data used as input for the "unit" tests</summary>

```log
***Begin monitoring DiskArbitration activity
***DiskAppeared ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:22:05.1438
***DiskAppeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1453
***DiskAppeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:05.1454
***DiskAppeared ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Time=20220108-20:22:05.1455
***DiskAppeared ('disk3', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1456
***DiskAppeared ((no BSD name), DAVolumePath = 'file:///System/Volumes/Data/home/', DAVolumeKind = 'autofs', DAVolumeName = '<null>') Time=20220108-20:22:05.1457
***DiskAppeared ('disk2s1', DAVolumePath = 'file:///Volumes/garuda-ext/', DAVolumeKind = 'hfs', DAVolumeName = 'garuda-ext') Time=20220108-20:22:05.1458
***DiskAppeared ('disk2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1459
***DiskAppeared ('disk0', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1460
***DiskAppeared ('disk0s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:05.1461
***DiskAppeared ('disk0s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1462
***DiskAppeared ('disk1', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.1463
***DiskAppeared ('disk1s1', DAVolumePath = 'file:///System/Volumes/Data/', DAVolumeKind = 'apfs', DAVolumeName = 'maindisk - Data') Time=20220108-20:22:05.1464
***DiskAppeared ('disk1s2', DAVolumePath = '<null>', DAVolumeKind = 'apfs', DAVolumeName = 'Preboot') Time=20220108-20:22:05.1465
***DiskAppeared ('disk1s3', DAVolumePath = 'file:///Volumes/Recovery/', DAVolumeKind = 'apfs', DAVolumeName = 'Recovery') Time=20220108-20:22:05.1466
***DiskAppeared ('disk1s4', DAVolumePath = 'file:///private/var/vm/', DAVolumeKind = 'apfs', DAVolumeName = 'VM') Time=20220108-20:22:05.1467
***DiskAppeared ('disk1s5', DAVolumePath = 'file:///', DAVolumeKind = 'apfs', DAVolumeName = 'maindisk') Time=20220108-20:22:05.1469
***DAIdle (no DADiskRef) Time=20220108-20:22:05.1470
***DiskUnmountApproval ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:22:09.9084
***DiskUnmountApproval ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:22:21.1909
***DiskDescriptionChanged ('disk4', DAVolumePath = '<null>') Time=20220108-20:22:21.5683
***DAIdle (no DADiskRef) Time=20220108-20:22:21.5684
***DiskDisappeared ('disk3', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:29.6767
***DiskDisappeared ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:22:29.6768
***DiskDisappeared ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Time=20220108-20:22:29.6770
***DiskDisappeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:29.6772
***DiskDisappeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:29.6773
***DAIdle (no DADiskRef) Time=20220108-20:22:29.6774
***DiskPeek ('disk3s1') Time=20220108-20:22:35.8607
***DiskAppeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:35.8673
***DiskMountApproval ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Comment=Approving Time=20220108-20:22:35.8686
***DiskPeek ('disk3s3') Time=20220108-20:22:36.0009
***DiskPeek ('disk3s2') Time=20220108-20:22:36.0011
***DiskPeek ('disk3') Time=20220108-20:22:36.0014
***DiskAppeared ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Time=20220108-20:22:36.0040
***DiskMountApproval ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Comment=Approving Time=20220108-20:22:36.0065
***DiskAppeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:36.0116
***DiskAppeared ('disk3', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:36.0118
***DAIdle (no DADiskRef) Time=20220108-20:22:36.0119
***DiskPeek ('disk4') Time=20220108-20:22:37.5920
***DiskAppeared ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:22:37.5980
***DiskMountApproval ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:22:37.5985
***DiskDescriptionChanged ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/') Time=20220108-20:22:41.5508
***DAIdle (no DADiskRef) Time=20220108-20:22:41.5509
***DiskUnmountApproval ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:22:58.3459
***DiskUnmountApproval ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:23:09.6402
***DiskDescriptionChanged ('disk4', DAVolumePath = '<null>') Time=20220108-20:23:10.0281
***DAIdle (no DADiskRef) Time=20220108-20:23:10.0282
***DiskDisappeared ('disk3', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:26:42.3640
***DiskDisappeared ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:26:42.3642
***DiskDisappeared ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Time=20220108-20:26:42.3643
***DiskDisappeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:26:42.3645
***DiskDisappeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:26:42.3647
***DAIdle (no DADiskRef) Time=20220108-20:26:42.3647
***DiskPeek ('disk3s1') Time=20220108-20:26:48.0983
***DiskAppeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:26:48.1052
***DiskMountApproval ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Comment=Approving Time=20220108-20:26:48.1069
***DiskPeek ('disk3s3') Time=20220108-20:26:48.2289
***DiskPeek ('disk3s2') Time=20220108-20:26:48.2291
***DiskPeek ('disk3') Time=20220108-20:26:48.2292
***DiskAppeared ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Time=20220108-20:26:48.2317
***DiskMountApproval ('disk3s3', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Boot OS X') Comment=Approving Time=20220108-20:26:48.2345
***DiskAppeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:26:48.2418
***DiskAppeared ('disk3', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:26:48.2419
***DAIdle (no DADiskRef) Time=20220108-20:26:48.2420
***DiskPeek ('disk4') Time=20220108-20:26:49.4535
***DiskAppeared ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:26:49.4590
***DiskMountApproval ('disk4', DAVolumePath = '<null>', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Comment=Approving Time=20220108-20:26:49.4594
***DiskDescriptionChanged ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/') Time=20220108-20:26:52.7814
***DAIdle (no DADiskRef) Time=20220108-20:26:52.7814

```
</details>


CHANGELOG.md
# 0.1.1

- Handle _Control-C_ in command-line utility.

# 0.1.0

- Initial version.


tests/std-stream-events.rs
#![allow(unused)]
use k9::assert_equal;
use mac_disk_monitor::event::Event;
use mac_disk_monitor::std::*;
use std::sync::mpsc::{channel, Receiver};
use std::thread;
use std::time::Duration;

#[test]
fn test_disk_activity() {
    let (action_sender, action_receiver) = channel();
    let (thread, receiver) =
        stream_events_with_command("./tests/dummy-disk-activity.sh", vec![], action_receiver);

    let event = receiver
        .recv_timeout(Duration::from_secs(5))
        .unwrap_or(None);
    assert_ne!(event, None);
    let event = event.unwrap();
    assert_equal!(event.name(), "DiskAppeared");
    assert_equal!(
        event.path().unwrap_or(String::new()),
        "file:///Volumes/my%20backups/"
    );

    action_sender.send(Action::Stop).unwrap();
    thread.join().unwrap().unwrap();
}

#[test]
fn test_timeout() {
    let (action_sender, action_receiver) = channel();
    let (thread, receiver) = stream_events_with_command("sleep", vec!["10"], action_receiver);

    let event = receiver.recv_timeout(Duration::from_secs(1)).unwrap_err();
    assert_equal!(event.to_string(), "timed out waiting on channel");

    action_sender.send(Action::Stop).unwrap();
    thread.join().unwrap().unwrap();
}

#[test]
fn test_stop_thread() {
    let (action_sender, action_receiver) = channel();
    let (thread, receiver) = stream_events_with_command("sleep", vec!["10"], action_receiver);

    action_sender.send(Action::Stop).unwrap();
    let event = receiver.recv_timeout(Duration::from_secs(1)).unwrap_err();
    assert_equal!(
        event.to_string(),
        "channel is empty and sending half is closed"
    );

    thread.join().unwrap().unwrap();
}


src/cli.rs
extern crate clap;
use clap::{App, Arg};

use ctrlc;
use mac_disk_monitor::std::*;
use mac_disk_monitor::version;
use std::sync::mpsc::channel;
use std::time::Duration;

/// Command-line entrypoint to monitor the disk events and print them out to stdout as json.
#[cfg(not(tarpaulin_include))]
fn main() {
    let version = version();
    let app = App::new("mac-disk-monitor")
        .version(version.as_str())
        .about("command-line tool to monitor disk activity on MacOS and output as json")
        .arg(
            Arg::with_name("format")
                .long("format")
                .help("the output format: (json|yaml)")
                .short("f")
                .default_value("json")
                .required(true)
                .takes_value(true),
        )
        .arg(
            Arg::with_name("interval")
                .long("interval")
                .help("how many milliseconds to wait for each event")
                .short("t")
                .default_value("3145")
                .required(true)
                .takes_value(true),
        );
    let matches = app.get_matches();

    let interval = matches
        .value_of("interval")
        .unwrap_or("")
        .parse::<u64>()
        .unwrap_or(3145);

    let format = matches.value_of("format").unwrap_or("json");
    match format {
        "json" | "yaml" => {}
        other => {
            eprintln!(
                "Invalid format {:?} should be either 'json' or 'yaml'",
                other
            );
            std::process::exit(1);
        }
    }

    let (action, receiver) = channel();
    let (thread, receiver) = stream_events(receiver);
    ctrlc::set_handler(move || {
        action
            .send(Action::Stop)
            .expect("Could not send 'stop' action to thread.");
    })
    .expect("Error setting Ctrl-C handler");

    loop {
        match receiver.recv_timeout(Duration::from_millis(interval)) {
            Ok(event) => match event {
                Some(event) => {
                    println!(
                        "{}",
                        match format {
                            "yaml" => event.to_yaml(),
                            _ => event.to_json(),
                        }
                    );
                }
                None => {
                    break;
                }
            },
            Err(e) => match e.to_string().as_str() {
                "timed out waiting on channel" => {}
                "channel is empty and sending half is closed" => {
                    break;
                }
                error => {
                    eprintln!("Error: {}", error);
                    break;
                }
            },
        }
    }
    eprintln!("waiting for thread to stop...");
    thread.join().unwrap().unwrap();
    eprintln!("done");
}


src/lib.rs
//! Mac Disk Monitor
//!
//! This crate provides a way to monitor disk activity on MacOS in real time.
//! Under the hood it simply calls `diskutil activity` in a thread and parses its output via [`Event::from_line()`]
//!
//!
//! Example usage:
//!
//! ```no_run
//! use mac_disk_monitor::{stream_events, Action};
//! use std::sync::mpsc::channel;
//! use std::time::Duration;
//!
//! fn main() {
//!     let (action, receiver) = channel();
//!     let (thread, receiver) = stream_events(receiver);
//!
//!     loop {
//!         // poll for events every 1.5 seconds
//!         match receiver.recv_timeout(Duration::from_millis(1500)) {
//!             Ok(event) => match event {
//!                 Some(event) => {
//!                     println!("{}", event.to_json());
//!                 }
//!                 None => {
//!                     eprintln!("thread has stopped...");
//!                 }
//!             },
//!             Err(e) => {
//!                 eprintln!("Error: {}", e);
//!                 break;
//!             }
//!         }
//!     }
//!     action.send(Action::Stop).unwrap();
//!     eprintln!("waiting for thread to stop...");
//!     thread.join().unwrap().unwrap()
//! }
//! ```
/// [`Event::from_line()`]: ./struct.Event.html#method.from_line
pub mod event;
pub mod std;
pub use crate::event::*;
pub use crate::std::*;

/// The version of the crate
pub fn version() -> String {
    let version = env!("CARGO_PKG_VERSION");
    String::from(version)
}


src/event.rs
use chrono::{Local, NaiveDateTime};
use regex::Regex;
use serde::{Deserialize, Serialize};

const APPLE_FORMAT: &'static str = "%Y%m%d-%H:%M:%S.%f";

/// Structured data about a disk event.
///
/// Can be either created empty or directly parsed from a string.
///
/// # Example:
///
/// ```
/// use mac_disk_monitor::Event;
///
/// let event = Event::from_line("***DiskAppeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:05.000001454");
/// assert_eq!(event.name(), "DiskAppeared");
/// assert_eq!(event.bsd_name().unwrap(), "disk3s1");
/// assert_eq!(event.kind().unwrap(), "msdos");
/// assert_eq!(event.volume_name().unwrap(), "EFI");
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Event {
    name: String,
    time: NaiveDateTime,
    bsd_name: Option<String>,
    volume_path: Option<String>,
    volume_kind: Option<String>,
    volume_name: Option<String>,
    comment: Option<String>,
}

impl Event {
    /// Creates an empty Event
    pub fn empty() -> Event {
        Event {
            name: String::new(),
            time: Local::now().naive_local(),
            bsd_name: None,
            volume_path: None,
            volume_kind: None,
            volume_name: None,
            comment: None,
        }
    }
    /// Creates a new Event containing the parsed the disk information from the given line.
    pub fn from_line(line: &str) -> Event {
        let mut event = Event::empty();
        //\s*\(('(?P<bsd_name>[^']+)')?, DAVolumePath\s*=\s*(?P<path>'[^']+')\)
        match extract_base_metadata(line) {
            Some((name, bsd_name, comment, time)) => {
                event.set_name(name.as_str());
                match bsd_name {
                    Some(bsd_name) => {
                        event.set_bsd_name(bsd_name.as_str());
                    }
                    None => {}
                }
                match comment {
                    Some(comment) => {
                        event.set_comment(comment.as_str());
                    }
                    None => {}
                }
                event.set_time_string(time.as_str());
            }
            None => {}
        }
        match extract_volume_path(line) {
            Some(path) => {
                event.set_path(path.as_str());
            }
            None => {}
        }
        match extract_volume_kind(line) {
            Some(kind) => {
                event.set_kind(kind.as_str());
            }
            None => {}
        }
        match extract_volume_name(line) {
            Some(name) => {
                event.set_volume_name(name.as_str());
            }
            None => {}
        }

        event
    }
    /// Serializes the Event to a yaml string
    pub fn to_yaml(&self) -> String
    where
        Self: Serialize,
    {
        match serde_yaml::to_string(&self) {
            Ok(val) => val,
            Err(e) => format!("{{\"error\": {:?}}}", e),
        }
    }
    /// Serializes the Event to a json string
    pub fn to_json(&self) -> String
    where
        Self: Serialize,
    {
        match serde_json::to_string(&self) {
            Ok(val) => val,
            Err(e) => format!("{}", e),
        }
    }

    /// The event name
    pub fn name(&self) -> String {
        self.name.clone()
    }

    /// Sets the event name
    pub fn set_name(&mut self, name: &str) {
        self.name = String::from(name)
    }
    /// Sets the bsd_name
    pub fn set_bsd_name(&mut self, bsd_name: &str) {
        self.bsd_name = Some(String::from(bsd_name));
    }
    /// The disk name, if any
    pub fn bsd_name(&self) -> Option<String> {
        self.bsd_name.clone()
    }
    /// Sets the event comment
    pub fn set_comment(&mut self, comment: &str) {
        self.comment = Some(String::from(comment));
    }
    /// The event comment, if any
    pub fn comment(&self) -> Option<String> {
        self.comment.clone()
    }

    /// Sets the volume path of the event
    pub fn set_path(&mut self, path: &str) {
        self.volume_path = if !path.eq("<null>") {
            Some(String::from(path))
        } else {
            None
        };
    }
    /// The volume path, if any
    pub fn path(&self) -> Option<String> {
        self.volume_path.clone()
    }

    /// Sets the volume kind, if any
    pub fn set_kind(&mut self, kind: &str) {
        self.volume_kind = if !kind.eq("<null>") {
            Some(String::from(kind))
        } else {
            None
        };
    }
    /// The volume kind, if any
    pub fn kind(&self) -> Option<String> {
        self.volume_kind.clone()
    }
    /// Sets the volume name
    pub fn set_volume_name(&mut self, name: &str) {
        self.volume_name = if !name.eq("<null>") {
            Some(String::from(name))
        } else {
            None
        };
    }
    /// The volume name, if any
    pub fn volume_name(&self) -> Option<String> {
        self.volume_name.clone()
    }
    pub fn set_time_string(&mut self, time: &str) {
        match NaiveDateTime::parse_from_str(time, APPLE_FORMAT) {
            Ok(time) => {
                self.time = time;
            }
            Err(e) => {
                panic!("Warning: {}", e);
            }
        }
    }
    /// The time when the event happened as string
    pub fn time_string(&self) -> String {
        self.time.format(APPLE_FORMAT).to_string()
    }
}

/// Extracts most of the metadata about the event from a line:
/// - event name,
/// - bsd_name
/// - comment
/// - time
pub fn extract_base_metadata(
    line: &str,
) -> Option<(String, Option<String>, Option<String>, String)> {
    let re =
        Regex::new(r"^[*]{3}(\w+)\s*\('?([^,']+)'?.*?\)\s*(Comment=(\S+))?\s*Time=(\S+)").unwrap();
    match re.captures(line) {
        Some(caps) => {
            let name = caps.get(1).unwrap().as_str().to_string();
            let bsd_name = caps.get(2).unwrap().as_str().to_string();
            let bsd_name = if bsd_name.eq("(no BSD name)") {
                None
            } else {
                Some(bsd_name)
            };
            let comment = match caps.get(4) {
                Some(m) => Some(m.as_str().to_string()),
                None => None,
            };
            let time = caps.get(5).unwrap().as_str().to_string();
            Some((name, bsd_name, comment, time))
        }
        None => None,
    }
}
/// Extracts the volume path from the given line.
pub fn extract_volume_path(line: &str) -> Option<String> {
    let re = Regex::new(r"DAVolumePath\s*=\s*('([^']+)')").unwrap();

    match re.captures(line) {
        Some(caps) => {
            let name = caps.get(2).unwrap().as_str().to_string();
            Some(name)
        }
        None => None,
    }
}
/// Extracts the volume kind from the given line.
pub fn extract_volume_kind(line: &str) -> Option<String> {
    let re = Regex::new(r"DAVolumeKind\s*=\s*('([^']+)')").unwrap();

    match re.captures(line) {
        Some(caps) => {
            let name = caps.get(2).unwrap().as_str().to_string();
            Some(name)
        }
        None => None,
    }
}
/// Extracts the volume name from the given line.
pub fn extract_volume_name(line: &str) -> Option<String> {
    let re = Regex::new(r"DAVolumeName\s*=\s*('([^']+)')").unwrap();

    match re.captures(line) {
        Some(caps) => {
            let name = caps.get(2).unwrap().as_str().to_string();
            Some(name)
        }
        None => None,
    }
}
#[cfg(test)]
mod tests {
    use super::{extract_base_metadata, Event};
    use k9::assert_equal;

    #[test]
    fn test_parse_disk_appeared_with_volume_path() {
        let line = String::from("***DiskAppeared ('disk4', DAVolumePath = 'file:///Volumes/my%20backups/', DAVolumeKind = 'hfs', DAVolumeName = 'Time Machine Backups') Time=20220108-20:22:05.000001438");
        let disk_appeared = Event::from_line(line.as_str());

        assert_equal!(disk_appeared.name().as_str(), "DiskAppeared");
        assert_equal!(disk_appeared.bsd_name(), Some(String::from("disk4")));
        assert_equal!(
            disk_appeared.path(),
            Some(String::from("file:///Volumes/my%20backups/"))
        );
        assert_equal!(disk_appeared.kind(), Some(String::from("hfs")));
        assert_equal!(
            disk_appeared.volume_name(),
            Some(String::from("Time Machine Backups"))
        );
        assert_equal!(
            disk_appeared.time_string().as_str(),
            "20220108-20:22:05.000001438"
        );
    }
    #[test]
    fn test_parse_disk_appeared_without_volume_path_kind_and_name() {
        let line = String::from("***DiskAppeared ('disk3s2', DAVolumePath = '<null>', DAVolumeKind = '<null>', DAVolumeName = '<null>') Time=20220108-20:22:05.000001453");
        let disk_appeared = Event::from_line(line.as_str());

        assert_equal!(disk_appeared.name().as_str(), "DiskAppeared");
        assert_equal!(disk_appeared.bsd_name(), Some(String::from("disk3s2")));
        assert_equal!(disk_appeared.path(), None);
        assert_equal!(disk_appeared.kind(), None);
        assert_equal!(disk_appeared.volume_name(), None);
        assert_equal!(
            disk_appeared.time_string().as_str(),
            "20220108-20:22:05.000001453"
        );
    }
    #[test]
    fn test_parse_disk_appeared_without_bsd_name() {
        let line = String::from("***DiskAppeared ((no BSD name), DAVolumePath = 'file:///System/Volumes/Data/home/', DAVolumeKind = 'autofs', DAVolumeName = '<null>') Time=20220108-20:22:05.000001457");
        let disk_appeared = Event::from_line(line.as_str());

        assert_equal!(disk_appeared.name().as_str(), "DiskAppeared");
        assert_equal!(disk_appeared.bsd_name(), None);
        assert_equal!(
            disk_appeared.path(),
            Some(String::from("file:///System/Volumes/Data/home/"))
        );
        assert_equal!(disk_appeared.kind(), Some(String::from("autofs")));
        assert_equal!(disk_appeared.volume_name(), None);
        assert_equal!(
            disk_appeared.time_string().as_str(),
            "20220108-20:22:05.000001457"
        );
    }
    #[test]
    fn test_parse_disk_disappeared() {
        let line = String::from("***DiskDisappeared ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Time=20220108-20:22:29.000006773");
        let event = Event::from_line(line.as_str());

        assert_equal!(event.name().as_str(), "DiskDisappeared");
        assert_equal!(event.bsd_name(), Some(String::from("disk3s1")));
        assert_equal!(event.path(), None);
        assert_equal!(event.kind(), Some(String::from("msdos")));
        assert_equal!(event.volume_name(), Some(String::from("EFI")));
        assert_equal!(event.time_string().as_str(), "20220108-20:22:29.000006773");
    }
    #[test]
    fn test_parse_disk_mount_approval() {
        let line = String::from("***DiskMountApproval ('disk3s1', DAVolumePath = '<null>', DAVolumeKind = 'msdos', DAVolumeName = 'EFI') Comment=Approving Time=20220108-20:22:35.000008686");
        let event = Event::from_line(line.as_str());

        assert_equal!(event.name().as_str(), "DiskMountApproval");
        assert_equal!(event.bsd_name(), Some(String::from("disk3s1")));
        assert_equal!(event.path(), None);
        assert_equal!(event.kind(), Some(String::from("msdos")));
        assert_equal!(event.volume_name(), Some(String::from("EFI")));
        assert_equal!(event.comment(), Some(String::from("Approving")));
        assert_equal!(event.time_string().as_str(), "20220108-20:22:35.000008686");
    }
    #[test]
    fn test_parse_disk_peek() {
        let line = String::from("***DiskPeek ('disk3s1') Time=20220108-20:22:35.000008607");
        let event = Event::from_line(line.as_str());

        assert_equal!(event.name().as_str(), "DiskPeek");
        assert_equal!(event.bsd_name(), Some(String::from("disk3s1")));
        assert_equal!(event.time_string().as_str(), "20220108-20:22:35.000008607");
    }
    #[test]
    fn test_parse_disk_activity_idle() {
        let line = String::from("***DAIdle (no DADiskRef) Time=20220108-20:22:29.000006774");
        let event = Event::from_line(line.as_str());

        assert_equal!(event.name().as_str(), "DAIdle");
        assert_equal!(event.bsd_name(), Some(String::from("no DADiskRef")));
        assert_equal!(event.time_string().as_str(), "20220108-20:22:29.000006774");
    }

    #[test]
    fn test_event_to_yaml() {
        let mut event = Event::empty();
        event.set_name("EventName");
        event.set_bsd_name("disk1337");
        event.set_volume_name("LeetDisk");
        event.set_path("file:///foo/bar");
        event.set_kind("stub");
        event.set_comment("dummy");
        event.set_time_string("20130307-13:37:42.1337");

        assert_equal!(
            event.to_yaml(),
            r#"---
name: EventName
time: "2013-03-07T13:37:42.000001337"
bsd_name: disk1337
volume_path: "file:///foo/bar"
volume_kind: stub
volume_name: LeetDisk
comment: dummy
"#
        );
    }
    #[test]
    fn test_event_to_json() {
        let mut event = Event::empty();
        event.set_name("EventName");
        event.set_bsd_name("disk1337");
        event.set_volume_name("LeetDisk");
        event.set_path("file:///foo/bar");
        event.set_kind("stub");
        event.set_comment("dummy");
        event.set_time_string("20130307-13:37:42.1337");

        assert_equal!(
            event.to_json(),
            r#"{"name":"EventName","time":"2013-03-07T13:37:42.000001337","bsd_name":"disk1337","volume_path":"file:///foo/bar","volume_kind":"stub","volume_name":"LeetDisk","comment":"dummy"}"#
        );
    }
    #[test]
    fn test_extract_base_metadata_invalid() {
        let parsed = extract_base_metadata("invalid event log");
        assert_equal!(parsed, None);
    }
}
